kyo-prelude/shared/src/main/scala/kyo2/kernel/RuntimeEffect.scala

```diff
@@ -0,0 +1,103 @@
+package kyo2.kernel
+
+import internal.*
+import kyo.Tag
+import kyo2.bug
+import kyo2.isNull
+import scala.util.NotGiven
+
+abstract class RuntimeEffect[+A]
+
+object RuntimeEffect:
+
+    inline def suspend[A, E <: RuntimeEffect[A]](inline tag: Tag[E]): A < E =
+        suspend(tag)(v => v)
+
+    inline def suspend[A, E <: RuntimeEffect[A], B, S](
+        inline tag: Tag[E]
+    )(
+        inline f: Runtime ?=> A => B < S
+    ): B < (E & S) =
+        suspend(tag, bug("Unexpected pending runtime effect: " + tag.show))(f)
+
+    inline def suspend[A, E <: RuntimeEffect[A]](
+        inline tag: Tag[E],
+        inline default: => A
+    ): A < Any =
+        suspend(tag, default)(v => v)
+
+    inline def suspend[A, E <: RuntimeEffect[A], B, S](
+        inline _tag: Tag[E],
+        inline default: => A
+    )(
+        inline f: Runtime ?=> A => B < S
+    )(using inline _frame: Frame): B < S =
+        new KyoDefer[B, S]:
+            def frame = _frame
+            def apply(v: Unit, values: Values)(using Runtime) =
+                Runtime.handle(
+                    suspend = this,
+                    continue = f(values.getOrElse(_tag, default).asInstanceOf[A])
+                )
+
+    inline def handle[A, E <: RuntimeEffect[A], B, S](
+        inline _tag: Tag[E],
+        inline ifUndefined: A,
+        inline ifDefined: A => A
+    )(v: B < (E & S))(
+        using inline _frame: Frame
+    ): B < S =
+        def handleLoop(v: B < (E & S))(using Runtime): B < S =
+            v match
+                case <(kyo: KyoSuspend[IX, OX, EX, Any, B, S] @unchecked) =>
+                    new KyoSuspend[IX, OX, EX, Any, B, S]:
+                        val tag   = kyo.tag
+                        val input = kyo.input
+                        def frame = _frame
+                        def apply(v: OX[Any], values: Values)(using Runtime) =
+                            val tag   = _tag
+                            val value = values.getOrElse(tag, null)
+                            val updated =
+                                if isNull(value) then values.set(tag, ifUndefined)
+                                else values.set(tag, ifDefined(value.asInstanceOf[A]))
+                            handleLoop(kyo(v, updated))
+                        end apply
+                case <(kyo) =>
+                    kyo.asInstanceOf[B]
+        handleLoop(v)
+    end handle
+
+    type WithoutRuntimeEffects[S] = S match
+        case RuntimeEffect[x] => Any
+        case s1 & s2          => WithoutRuntimeEffects[s1] & WithoutRuntimeEffects[s2]
+        case _                => S
+
+    def boundary[A, B, S, S2](v: A < S)(
```

Comment:
@johnhungerford I'm sorry for asking for your help again with these type-level issues 😅 but I'm having trouble using this new boundary method. The idea is to use it when forking fibers:

```scala
object Fiber:
    // I'd need to restrict `S` here only to runtime effects
    def init[A, S](v: A < (S & Fibers)): Fiber[A] < (IO & S) =
        Runtime.boundary(v) { (cont: A < Fibers) =>
            // schedule execution without `S` pending
        }
```

I've tried a few ways, including the `Has` and `Reducible` mechanisms, but had no luck. Using type constraints like `_ <: RuntimeEffect[?]` also doesn't work because `(SomeRuntimeEffect & NonRuntimeEffect) <:< RuntimeEffect[?]`.

Reply:
I'll take a look at this tomorrow. This might be tricky though. I haven't had much luck using type class derivation or inlining to iterate over the components of an intersection type.